### ISSUE TITLE
build broken with clang 5.0 and Apple's clang 9.0.0

### ISSUE THREAD

**Issue by davydden** [dealii/dealii#5770]
subj

```
Undefined symbols for architecture x86_64:
  "dealii::GeometryInfo<0>::ucd_to_deal", referenced from:
      std::__1::basic_istream<char, std::__1::char_traits<char> >& dealii::DataOutBase::operator>><0, 1>(std::__1::basic_istream<char, std::__1::char_traits<char> >&, dealii::DataOutBase::Patch<0, 1>&) in data_out_base.cc.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& dealii::DataOutBase::operator<<<0, 1>(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, dealii::DataOutBase::Patch<0, 1> const&) in data_out_base.cc.o
      std::__1::basic_istream<char, std::__1::char_traits<char> >& dealii::DataOutBase::operator>><0, 2>(std::__1::basic_istream<char, std::__1::char_traits<char> >&, dealii::DataOutBase::Patch<0, 2>&) in data_out_base.cc.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& dealii::DataOutBase::operator<<<0, 2>(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, dealii::DataOutBase::Patch<0, 2> const&) in data_out_base.cc.o
      std::__1::basic_istream<char, std::__1::char_traits<char> >& dealii::DataOutBase::operator>><0, 3>(std::__1::basic_istream<char, std::__1::char_traits<char> >&, dealii::DataOutBase::Patch<0, 3>&) in data_out_base.cc.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& dealii::DataOutBase::operator<<<0, 3>(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, dealii::DataOutBase::Patch<0, 3> const&) in data_out_base.cc.o
ld: symbol(s) not found for architecture x86_64
```

most likely broken by ~~https://github.com/dealii/dealii/pull/5748~~ https://github.com/dealii/dealii/pull/5745/

@masterleinad any ideas?

**Comment by davydden** [comment]
p.s. I looked at the `geometry_info.h` and I don't know what's going on...

**Comment by jppelteret** [comment]
I feel like I might have had this problem when putting together the `ad_number_traits` - using `constexpr` also lead to linking errors in some of the tests I made, but only when using `clang`. See [this comment](https://github.com/dealii/dealii/blob/master/include/deal.II/differentiation/ad/ad_number_traits.h#L757) and what follows for my workaround. Maybe we need to do similar for `GeometryInfo.`

**Comment by davydden** [comment]
thanks @jppelteret , i will give it a quick look. I was just suprized why it complains about such things only for `GeometryInfo<0>::ucd_to_deal`? There whole lot of other variables and dimension and that seemed to work. Mystery if you ask me...

**Comment by davydden** [comment]
Looks like it's related to [odr-uses](http://en.cppreference.com/w/cpp/language/definition#ODR-use), see [stackoverflow's answer](https://stackoverflow.com/a/28846608/888478).

**Comment by masterleinad** [comment]
The definition is currently in for these variables is currently `geometry_info.cc`. Does moving
```
constexpr std::array<unsigned int, GeometryInfo<0>::vertices_per_cell>
GeometryInfo<0>::ucd_to_deal;

constexpr std::array<unsigned int, GeometryInfo<0>::vertices_per_cell>
GeometryInfo<0>::dx_to_deal;
```
to `geometry_info.h` fix this for you?

**Comment by davydden** [comment]
i wil try, but first i will try adding `template struct GeometryInfo<0>;`

**Comment by bangerth** [comment]
What if you change all occurrences of `constexpr` in the .h/.cc file to `const`? For what we use it for, it will not make a difference, but it would be interesting to see whether the compiler likes it better. If so, it may simply be a compiler bug. If it's the same result as before, then it's likely a problem in our code.

**Comment by davydden** [comment]
@dealii/dealii any1 with a macOS can reproduce the issue?

p.s. I have

```
$ clang --version
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin17.3.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by drwells** [comment]
@davydden I do not have macOS but compiling on my machine with clang 5.0 results in link failures:

```
../lib/libdeal_II.g.so.9.0.0-pre: error: undefined reference to 'dealii::GeometryInfo<0>::dx_to_deal'
../lib/libdeal_II.g.so.9.0.0-pre: error: undefined reference to 'dealii::GeometryInfo<0>::ucd_to_deal'
clang-5.0: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [examples/CMakeFiles/step-32.debug.dir/build.make:233: bin/step-32.debug] Error 1
```

**Comment by davydden** [comment]
thanks @drwells for confirming.

**Comment by davydden** [comment]
What I find strange is that we only have issues with dim=0 where explicit specialization is used. Could be a hint towards the solution.

**Comment by luca-heltai** [comment]
I can confirm on my mac:

~~~
Undefined symbols for architecture x86_64:
  "dealii::GeometryInfo<0>::ucd_to_deal", referenced from:
      std::__1::basic_istream<char, std::__1::char_traits<char> >& dealii::DataOutBase::operator>><0, 1>(std::__1::basic_istream<char, std::__1::char_traits<char> >&, dealii::DataOutBase::Patch<0, 1>&) in data_out_base.cc.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& dealii::DataOutBase::operator<<<0, 1>(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, dealii::DataOutBase::Patch<0, 1> const&) in data_out_base.cc.o
      std::__1::basic_istream<char, std::__1::char_traits<char> >& dealii::DataOutBase::operator>><0, 2>(std::__1::basic_istream<char, std::__1::char_traits<char> >&, dealii::DataOutBase::Patch<0, 2>&) in data_out_base.cc.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& dealii::DataOutBase::operator<<<0, 2>(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, dealii::DataOutBase::Patch<0, 2> const&) in data_out_base.cc.o
      std::__1::basic_istream<char, std::__1::char_traits<char> >& dealii::DataOutBase::operator>><0, 3>(std::__1::basic_istream<char, std::__1::char_traits<char> >&, dealii::DataOutBase::Patch<0, 3>&) in data_out_base.cc.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& dealii::DataOutBase::operator<<<0, 3>(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, dealii::DataOutBase::Patch<0, 3> const&) in data_out_base.cc.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
~~~

**Comment by masterleinad** [comment]
Going back to `const` fixes this for me using `clang-5.0.0`. PR in a minute.
